versions/next/asyncapi.md

```diff
@@ -2208,6 +2231,68 @@ petstore_auth:
 - read:pets
 ```
 
+### <a name="correlationIdObject"></a>Correlation ID Object
+
+An object that specifies an identifier at design time that can used for message tracing and correlation.
+
+For specifying and computing the location of a Correlation ID a [runtime expression](#runtimeExpression) is used.
+
+##### Fixed Fields
+
+Field Name | Type | Description
+---|:---|---
+description | `string` | A optional description of the identifier. [CommonMark syntax](http://spec.commonmark.org/) can be used for rich text representation.
+type | `integer` \| `string` \| `number` | An optional type specification.
+location | {expression} | **REQUIRED.** A runtime expression that specifies the location of the correlation ID.
+
+##### Examples
+
+```json
+{
+  "description": "Default Correlation ID",
+  "type": "string",
+  "location": "$message.header.correlationId"
+}
+```
+
+```yaml
+description: Default Correlation iD
+type: string
+location: $message.header.correlationId
+```
+
+### <a name="runtimeExpression"></a>Runtime Expression
+
+A runtime expression allows values to be defined based on information that will be available within the message.
+This mechanism is used by [Correlation ID Object](#correlationIdObject).
+
+The runtime expression is defined by the following [ABNF](https://tools.ietf.org/html/rfc5234) syntax
+
+```
+      expression = ( "$message" )
+      source = ( header-reference | payload-reference )  
+      header-reference = "header." token
+      payload-reference = "payload." name
+      fragment = a JSON Pointer [RFC 6901](https://tools.ietf.org/html/rfc6901) 
+      name = *( char )
+      char = as per RFC [7159](https://tools.ietf.org/html/rfc7159#section-7)
+      token = as per RFC [7230](https://tools.ietf.org/html/rfc7230#section-3.2.6)
+```
+
+The `name` identifier is case-sensitive, whereas `token` is not. 
+
+The table below provides examples of runtime expressions and examples of their use in a value:
+
+##### <a name="runtimeExpressionExamples"></a>Examples
+
+Source Location | Example expression  | Notes
+---|:---|:---|
+Message Header Property | `$message.header.correlationId` | Headers are analogous to message properties
+Message Payload Property | `$message.payload.meta.messageId` | Correlation ID is set using a property from the `meta` section of a message
+
+Runtime expressions preserve the type of the referenced value.
+Expressions can be embedded into string values by surrounding the expression with `{}` curly braces.
```

Comment:
Do you have an example where this should be possible?

Reply:
Where you use a custom correlation ID? This would be down to API designers - if they want to enforce a correlation ID and cannot add a message header property then this offers the requisite features to use something in the payload. Alternatively, they may be retrofitting an API specification to a messaging platform and already use a custom correlation ID in the message payload.

Is this what you were getting at?

Reply:
You mean the curly braces? Will add an example

Reply:
I'm more worried about where you can use runtime expressions. Specification-wise it's easy to say we support them, but tooling-wise it's not that simple. I want to make sure we only support things where we really need them.

Reply:
On reflection we don't support this at the moment. In OpenAPI it's a construct that relates to [Callback Object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#callbackObject). I can't see a good fit so will remove the guidance.

Reply:
> I'm more worried about where you can use runtime expressions. Specification-wise it's easy to say we support them, but tooling-wise it's not that simple. I want to make sure we only support things where we really need them.

I think the runtime expression support is a really good investment for the future. If you look at the way Callback Object in OpenAPI can now be engineered it adds a great deal of flexibility for the API designer. Specifically, when we tackle RPC in earnest a runtime expression to specify replyTo queues and the like based on attributes of the message will give API designers a real boost IMHO.

Reply:
Yeah, I get it. But I still can't see the value outside the correlation ID feature or a possible RPC support in the future. My concern is just that this sentence really looked like you can embed runtime expressions anywhere you have a string. Anyway, I think your previous comment clarified it :)

Reply:
👍